src/lib/io/sol-iio.c

```diff
@@ -963,9 +1052,31 @@ sol_iio_close(struct sol_iio_device *device)
 {
     int i;
     struct sol_iio_channel *channel;
+    struct iio_opened_device *entry = NULL;
+    struct iio_opened_device *remove_entry = NULL;
 
     SOL_NULL_CHECK(device);
 
+    SOL_VECTOR_FOREACH_IDX (&iio_opened_devices, entry, i) {
+        if (device->device_id == entry->device->device_id) {
+            entry->refcount--;
+            if (entry->refcount == 0) {
+                SOL_DBG("Close device%d", device->device_id);
+                remove_entry = entry;
+                free(entry->config);
+            } else {
+                SOL_DBG("Won't close device%d refcount=%d > 0", device->device_id, entry->refcount);
+                return;
+            }
```

Comment:
Minor: you can `break` at the end of the `if (device->...` to save some iterations.